### ISSUE TITLE
JSON Errors After Upgrading MkDocs, Unable to Serve or Build

### ISSUE THREAD

**Issue by chetcuti** [mkdocs/mkdocs#1650]
I finally upgraded MkDocs from 0.16.0 to the newest version today, and although I believe I've updated the mkdocs.yml file accordingly (such as changing theme_dir to custom_dir, etc.), I'm getting errors when I try to serve or build the documentation. I'm running Python 2.7.15 and MkDocs 1.0.4.

This is what I get when I try to run 'mkdocs build' from the documentation directory (which seems to be about the same as I get when I try to run 'mkdocs serve').

Any idea what might be going on here?

```
INFO    -  Cleaning site directory
INFO    -  Building documentation to directory: x:\dev\domainmod.userguide\site
ERROR   -  Error building page 'index.md': Undefined is not JSON serializable
Traceback (most recent call last):
  File "c:\python27\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\python27\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Python27\Scripts\mkdocs.exe\__main__.py", line 9, in <module>
  File "c:\python27\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "c:\python27\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "c:\python27\lib\site-packages\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\python27\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\python27\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "c:\python27\lib\site-packages\mkdocs\__main__.py", line 163, in build_command
    ), dirty=not clean)
  File "c:\python27\lib\site-packages\mkdocs\commands\build.py", line 295, in build
    _build_page(file.page, config, files, nav, env, dirty)
  File "c:\python27\lib\site-packages\mkdocs\commands\build.py", line 216, in _build_page
    output = template.render(context)
  File "c:\python27\lib\site-packages\jinja2\environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "c:\python27\lib\site-packages\jinja2\environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "x:\dev\domainmod.userguide\theme/main.html", line 1, in top-level template code
    {% extends "base.html" %}
  File "x:\dev\domainmod.userguide\theme/base.html", line 30, in top-level template code
    {%- block libs %}
  File "x:\dev\domainmod.userguide\theme/base.html", line 35, in block "libs"
    var mkdocs_page_input_path = {{ page.input_path|tojson|safe }};
  File "c:\python27\lib\site-packages\mkdocs\utils\filters.py", line 8, in tojson
    return jinja2.Markup(json.dumps(obj, **kwargs))
  File "c:\python27\lib\json\__init__.py", line 244, in dumps
    return _default_encoder.encode(obj)
  File "c:\python27\lib\json\encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "c:\python27\lib\json\encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "c:\python27\lib\json\encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: Undefined is not JSON serializable
```

**Comment by waylan** [comment]
Note this line of your traceback:

```
 File "x:\dev\domainmod.userguide\theme/base.html", line 35, in block "libs"
    var mkdocs_page_input_path = {{ page.input_path|tojson|safe }};
```

That is were the error is coming from. Normally, I would suggest reporting the problem to the theme devs, but this appears to be your won custom template. The error is reporting that `page.input_path` is "Undefined", and of course, you cannot JSON serialize and undefined variable. Generally, when a template reports that a variable is undefined, the problem is that the parent variable is `None`. Of course, for Markdown pages, the `page` variable contains the `Page` object. But for none-Markdown pages (such as static template pages), the `page` variable is `None` and so `None.input_path` would be undefined and raise an error.

That is why the built-in themes always check that `page` is not `None` before attempting to access `page` or any of its child attributes.

**Comment by chetcuti** [comment]
Thank you so much for pointing me in the right direction! It took a bit of time but I'm all upgraded now and everything's working great.

To anyone reading this that runs into the same problem, as @waylan pointed out, this is an issue with the theme. Originally I had copied the readthedocs theme and made a few cosmetic changes to create my own custom theme, and since some of the variables and how they work have changed since the old version of MkDocs (as well the older versions of the default themes), the theme was now broken.

To fix this I used [WinMerge](http://winmerge.org) to compare the new version of the readthedocs theme and the old version that I had customized, and I then re-copied my customizations over to the new version of the theme, one-by-one. It wasn't fun, but it's done, and now I have the newest version of MkDocs.
